### ISSUE TITLE
Incompatible with SSR (since v1.4.0)

### ISSUE THREAD

**Issue by pilwon** [jamesmfriedman/rmwc#169]
The latest version (v1.4.0) introduced new incompatibility issue with server-side rendering (SSR):

```
ReferenceError: HTMLElement is not defined
    at Function.createAdapter (./node_modules/@material/ripple/dist/mdc.ripple.js:858:79)
    at Ripple.getDefaultFoundation (./node_modules/@material/ripple/dist/mdc.ripple.js:813:89)
    at Ripple.Foundation (./node_modules/rmwc/Base/MDCFoundation.js:134:33)
    at new Ripple (./node_modules/rmwc/Ripple/index.js:44:97)
```

It seems [this code](https://github.com/stasson/vue-mdc-adapter/blob/master/components/ripple/mdc-ripple-base.js) from [vue-mdc-adapter](https://github.com/stasson/vue-mdc-adapter) project already figured out the solution for the same problem.

**Comment by jamesmfriedman** [comment]
Thanks for reporting. I’ve never actually tested this with SSR. I’ll look into building a test environment and troubleshooting this. 

Until then, I would say stay on 1.3.x

**Comment by jamesmfriedman** [comment]
@pilwon Got a fix in place.

Can you do me a favor and please verify that this PR fixes your issue? If so, I can put out a patch release today.
https://github.com/jamesmfriedman/rmwc/pull/170

**Comment by jrajav** [comment]
@jamesmfriedman I see that this fix has something to do with MutationObserver, just wanted to note that 1.4.0 seems to break my Jest tests as well:

```
FAIL  src/components/browse/BrowsePage.test.tsx
  ● Test suite failed to run

    Cannot find module 'mutation-observer' from 'testPolyfill.js'

      at Resolver.resolveModule (node_modules/jest-runner/node_modules/jest-resolve/build/index.js:194:17)
      at Object.<anonymous> (node_modules/rmwc/Base/testPolyfill.js:7:25)
```

(Running on a slightly old version for various reasons, Jest 21.2.1 and jest-enzyme 4.0.2, Babel 7)

**Comment by jamesmfriedman** [comment]
@jrajav was that error on the PR branch or 1.4.0?

**Comment by jamesmfriedman** [comment]
@jrajav I just double checked https://github.com/jamesmfriedman/rmwc/pull/170. The dev for the mutation observer is in there. It's one of the polyfills needed for the material library that they just started using.

The error that you're getting sounds like the dep is just missing from you node modules. I'd suggest trying to clear your npm cache and reinstalling. 

@pilwon issue was related to timing of when the MDCFoundation constructors were being called (in the constructor instead of componentDidMount).

It would  be great to get validation from both of you on the PR since I'm not experiencing either issue and I want to make sure it's actually solving your problem.

**Comment by pilwon** [comment]
@jamesmfriedman I tested PR and it worked. Thanks for the prompt fix for SSR 👍

**Comment by jamesmfriedman** [comment]
That’s great news @pilwon. From here on out I’ll make sure I add tests for basic ssr on all components. 

@jrajav I’m still convinced your node modules just need to be reinstallled. Out of town for the weekend but I’ll get this in a patch release for Monday morning.

**Comment by jrajav** [comment]
@jamesmfriedman It still fails, when I just try to install that PR branch as a git npm dependency:

<img width="928" alt="screen shot 2018-03-25 at 10 25 11" src="https://user-images.githubusercontent.com/2074135/37876794-605f64c8-3017-11e8-8613-38af96085de8.png">

Here's my package.json:

```
{
  "name": "jrajav-repo",
  "version": "0.1.0",
  "description": "description",
  "main": "webpack.config.js",
  "private": true,
  "scripts": {
    "build:webpack": "webpack",
    "build": "yarn run clean && yarn run build:webpack",
    "checkdeps": "check-dependencies",
    "clean": "rimraf dist",
    "eslint": "eslint src --ext=ts --ext=tsx --ext=json",
    "lint": "yarn run checkdeps && yarn run eslint && yarn run stylelint",
    "lint:fix": "eslint src --ext=ts --ext=tsx --ext=json --fix",
    "notify:pass": "notify -t \"Redub frontend: Verify\" -m \"✅ All verifications passed\"",
    "notify:fail": "notify -t \"Redub frontend: Verify\" -m \"❌ Some verifications failed\"",
    "start": "webpack-dev-server",
    "start:dist": "yarn run build && http-server ./dist -p 3000",
    "stylelint": "stylelint \"src/**/*.sss\"",
    "test": "jest --config jest.config.json",
    "test:watch": "yarn run test --watch",
    "test:coverage": "yarn run test --coverage && opn coverage/lcov-report/index.html",
    "ts": "tsc --pretty",
    "verify:lint": "yarn run lint",
    "verify:ts": "yarn run ts",
    "verify:test": "yarn run test",
    "verify": "( run-p verify:* && yarn run notify:pass ) || yarn run notify:fail",
    "watch": "cross-env NO_UPDATE_NOTIFIER=true npm-watch"
  },
  "moduleRoots": [
    "src"
  ],
  "watch": {
    "lint": {
      "patterns": [
        "src"
      ],
      "extensions": "js,jsx"
    },
    "test": {
      "patterns": [
        "src"
      ],
      "extensions": "js,jsx"
    }
  },
  "nyc": {
    "extension": [
      ".js",
      ".jsx"
    ],
    "exclude": [
      "test-setup.js",
      "**/*.spec.{js,jsx}"
    ],
    "sourceMap": false,
    "instrument": false
  },
  "author": "Jonathan Rajavuori <jrajav@gmail.com>",
  "repository": {
    "type": "git",
    "url": "https://gitlab.com/dubtrack.fm/re/frontend"
  },
  "license": "MIT",
  "dependencies": {
    "@babel/cli": "^7.0.0-beta.40",
    "@babel/core": "^7.0.0-beta.40",
    "@babel/plugin-proposal-class-properties": "^7.0.0-beta.40",
    "@babel/plugin-transform-block-scoping": "^7.0.0-beta.40",
    "@babel/polyfill": "^7.0.0-beta.40",
    "@babel/preset-env": "^7.0.0-beta.40",
    "@babel/preset-react": "^7.0.0-beta.40",
    "@babel/preset-typescript": "^7.0.0-beta.40",
    "@babel/register": "^7.0.0-beta.40",
    "autodll-webpack-plugin": "^0.3.8",
    "axios": "^0.17.1",
    "babel-core": "^7.0.0-beta.40",
    "babel-jest": "^22.0.6",
    "babel-loader": "8.0.0-beta.0",
    "babel-plugin-ramda": "^1.4.3",
    "babel-plugin-react-css-modules": "github:jrajav/babel-plugin-react-css-modules#release",
    "babel-plugin-root-import": "^5.1.0",
    "big-integer": "^1.6.26",
    "case-sensitive-paths-webpack-plugin": "^2.1.1",
    "css-loader": "^0.28.9",
    "extract-text-webpack-plugin": "^3.0.2",
    "file-loader": "^1.1.8",
    "firebase": "^4.10.0",
    "fuzzysort": "^1.0.5",
    "html-entities": "^1.2.1",
    "html-webpack-plugin": "^2.30.1",
    "i18next": "^10.4.1",
    "i18next-browser-languagedetector": "^2.1.0",
    "i18next-resource-store-loader": "^0.1.2",
    "immutable": "^3.8.2",
    "immutable-to-js": "^1.0.0",
    "jest-cli": "^22.0.6",
    "js-yaml": "^3.10.0",
    "lodash.throttle": "^4.1.1",
    "material-components-web": "^0.33.0",
    "mitt": "^1.1.3",
    "moment": "^2.20.1",
    "moment-duration-format": "^2.2.2",
    "node-sass": "^4.7.2",
    "normalize.css": "^8.0.0",
    "page": "^1.8.3",
    "postcss-cssnext": "^3.1.0",
    "postcss-import-sync2": "^1.0.0",
    "postcss-loader": "^2.1.0",
    "qs": "^6.5.1",
    "ramda": "^0.25.0",
    "react": "^16.2.0",
    "react-dom": "^16.2.0",
    "react-i18next": "^7.4.0",
    "react-image": "^1.3.1",
    "react-player": "^1.1.2",
    "react-redux": "^5.0.7",
    "react-redux-firebase": "^2.0.4",
    "react-sortable-hoc": "^0.6.8",
    "react-svg-loader": "^2.1.0",
    "react-timeout": "^1.0.1",
    "react-timer-hoc": "^2.3.0",
    "react-virtualized": "^9.18.5",
    "react-youtube": "^7.5.0",
    "recompose": "^0.26.0",
    "redux": "^3.7.2",
    "redux-actions": "^2.2.1",
    "redux-devtools-extension": "^2.13.2",
    "redux-immutable": "^4.0.0",
    "redux-promise-middleware": "^5.0.0",
    "redux-thunk": "^2.2.0",
    "rimraf": "^2.6.2",
    "rmwc": "git://github.com/jamesmfriedman/rmwc.git#ssr",
    "sass-loader": "^6.0.6",
    "socket.io-client": "^2.0.4",
    "style-loader": "^0.20.2",
    "sugarss": "^1.0.1",
    "type-to-reducer": "^1.0.4",
    "webpack": "^3.11.0",
    "webpack-merge": "^4.1.1"
  },
  "devDependencies": {
    "@firebase/app-types": "^0.1.1",
    "@types/enzyme": "^3.1.9",
    "@types/jest": "^22.1.3",
    "@types/moment-duration-format": "^2.2.0",
    "@types/node": "^9.4.6",
    "@types/page": "^1.5.32",
    "@types/react": "^16.0.38",
    "@types/react-css-modules": "^4.6.1",
    "@types/react-dom": "^16.0.4",
    "@types/redux-actions": "^2.2.4",
    "@types/socket.io-client": "^1.4.32",
    "@types/webpack": "^3.8.8",
    "@types/webpack-env": "^1.13.5",
    "babel-eslint": "^8.2.2",
    "check-dependencies": "^1.1.0",
    "convict": "^4.0.2",
    "cross-env": "^5.1.3",
    "enzyme": "^3.3.0",
    "enzyme-adapter-react-16": "^1.1.1",
    "enzyme-to-json": "^3.3.1",
    "eslint": "^4.18.1",
    "eslint-plugin-fp": "^2.3.0",
    "eslint-plugin-html": "^4.0.2",
    "eslint-plugin-json": "^1.2.0",
    "eslint-plugin-react": "^7.7.0",
    "exports-loader": "^0.7.0",
    "favicons-webpack-plugin": "0.0.7",
    "http-server": "^0.11.1",
    "identity-obj-proxy": "^3.0.0",
    "jest": "^21.2.1",
    "jest-css-modules": "^1.1.0",
    "jest-enzyme": "^4.0.2",
    "jsdom": "^11.6.2",
    "loglevel": "^1.6.1",
    "node-notifier-cli": "^1.1.2",
    "npm-run-all": "^4.1.2",
    "npm-watch": "^0.3.0",
    "opn-cli": "^3.1.0",
    "redux-jest": "^1.1.1",
    "stylelint": "^9.0.0",
    "stylelint-config-standard": "^18.1.0",
    "typescript": "^2.7.2",
    "webpack-dev-server": "^2.11.1"
  }
}
```

It also fails with different errors when I try to clone and link rmwc, but I think that's just because I don't know how to build it properly - what's the right script to run for that?

**Comment by jrajav** [comment]
By the way, this was after several cycles of deleting my node_modules directory and re-installing everything.

**Comment by jamesmfriedman** [comment]
Sorry for the hiccup @jrajav. Still don't know the exact issue you're having.

I will say that if you cloned and linked the repo, then you should run `npm run babelfy` from the rmwc directory. That will spit out the "dist" version. I need to document how to link the module locally and probably just add that as part of the build script.

It would be helpful to see a minimal repro, or I'd be happy to jump on a skype call and see what the exact issue is.

**Comment by jamesmfriedman** [comment]
@pilwon this is fixed in 1.4.1.

@jrajav, jump on Gitter and let me know if that last tip helped for building locally. I've added a task to document this, but I'll also see if I can make it automatic with npm postinstall.
